py/dml/codegen.py

```diff
@@ -394,6 +395,462 @@ def memoization_common_prelude(name, site, outp, throws, make_ref):
+
+
+class AfterOnHookIntoMethodInfo(AfterOnHookInfo):
+    def __init__(self, typeseq_info, method, param_to_msg_comp):
+        self.method = method
+        super().__init__(method.dimsizes, method.parent, typeseq_info, method,
+                         param_to_msg_comp, method.inp, bool(self.method.inp))
+        self._args_type = (
+            TStruct({name: typ
+                    for (i, (name, typ)) in enumerate(method.inp)
+                     if i not in param_to_msg_comp},
+                    label=f'_after_on_hook_{self.uniq}_args')
+            if len(self.method.inp) > len(param_to_msg_comp) else None)
+
+    def generate_callback_call(self, indices_lit, args_lit, msg_lit):
+        site = self.method.site
+        indices = tuple(mkLit(site, f'{indices_lit}[{i}]', TInt(32, False))
+                        for i in range(self.method.dimensions))
+        args = tuple(
+            mkLit(site,
+                  f'{msg_lit}->comp{self.param_to_msg_comp[i]}'
+                  if i in self.param_to_msg_comp else f'{args_lit}->{pname}',
+                  ptype)
+            for (i, (pname, ptype)) in enumerate(self.method.inp))
+        with LogFailure(site, self.method, indices), \
+             crep.DeviceInstanceContext():
+            code = codegen_call(site, self.method, indices, args, ())
+        code = mkCompound(site, [code])
+        code.toc()
+
+    def generate_args_serializer(self, site, args_expr, out_expr):
+        sources = tuple((ctree.mkSubRef(site, args_expr, name, "."),
+                         safe_realtype(typ))
+                        if i not in self.param_to_msg_comp else None
+                        for (i, (name, typ)) in enumerate(self.method.inp))
+        serialize.serialize_sources_to_list(site, sources, out_expr)
+
+    def generate_args_deserializer(self, site, val_expr, out_expr, error_out):
+        if self.args_type:
+            tmp_out_decl, tmp_out_ref = serialize.declare_variable(
+                site, '_tmp_out', self.args_type)
+            pre = [tmp_out_decl]
+            post = [ctree.mkAssignStatement(site, out_expr,
+                                            ctree.ExpressionInitializer(
+                                                tmp_out_ref))]
+        else:
+            tmp_out_ref = None
+            pre = []
+            post = []
+        targets = tuple((ctree.mkSubRef(site, tmp_out_ref, name, "."),
+                         safe_realtype(typ))
+                        if i not in self.param_to_msg_comp else None
+                        for (i, (name, typ)) in enumerate(self.method.inp))
+
+        serialize.deserialize_list_to_targets(
+            site, val_expr, targets, error_out,
+            f'deserialization of arguments to {self.method.name}',
+            pre, post)
+
+    @property
+    def args_type(self):
+        return self._args_type
+
+    @property
+    def types_to_declare(self):
+        return (self._args_type,) * (self._args_type is not None)
```

Comment:
I geekily wonder if this can be expressed
```
@property
def types_to_declare(self):
  if self._args_type  is not None:
    yield self._args_type
```
If so, I don't have an opinion on which one is better.

Reply:
It's certainly not the same semantically. One returns a generator, the other doesn't, so e.g. with your definition `bool(obj.types_to_declare)` will always return `True`.

Reply:
so depends on whether you write a docstring or type annotation that specifically prescribes a tuple, or if it says an iterable is enough.

Reply:
Funny you should say that! I did write a docstring for `types_to_declare` without reading this latest comment, and in it I indeed state "iterable"! So yeah, either works.